python_modules/dagster/dagster/_core/execution/plan/execute_step.py

```diff
@@ -779,15 +780,32 @@ def _gen_fn():
 
     asset_key, partitions = _asset_key_and_partitions_for_output(output_context)
     if asset_key:
-        for materialization in _get_output_asset_materializations(
-            asset_key,
-            partitions,
-            output,
-            output_def,
-            manager_metadata,
-            step_context,
-        ):
-            yield DagsterEvent.asset_materialization(step_context, materialization)
+        asset_layer = step_context.job_def.asset_layer
+        execution_type = (
+            asset_layer.assets_def_for_asset(asset_key).asset_execution_type_for_asset(asset_key)
+            if asset_layer.has_assets_def_for_asset(asset_key)
+            else AssetExecutionType.MATERIALIZATION
+        )
+
+        check.invariant(
+            execution_type != AssetExecutionType.UNEXECUTABLE,
+            "There should never be unexecutable assets here",
+        )
+        if execution_type == AssetExecutionType.MATERIALIZATION:
+            for materialization in _get_output_asset_materializations(
+                asset_key,
+                partitions,
+                output,
+                output_def,
+                manager_metadata,
+                step_context,
+            ):
+                yield DagsterEvent.asset_materialization(step_context, materialization)
+        elif execution_type == AssetExecutionType.OBSERVATION:
+            # if this is an observation execution type, yield no materializations
+            pass
+        else:
+            check.failed(f"Unexpected asset execution type {execution_type}")
```

Comment:
hmm not sure if theres a way to make this feel cleaner, take another second to stare at it and see if you can come up with something